reconcile/skupper_network/integration.py

```diff
@@ -0,0 +1,282 @@
+import logging
+import sys
+from collections.abc import (
+    Callable,
+    Iterable,
+)
+from typing import (
+    Any,
+    Optional,
+)
+
+from sretoolbox.utils import threaded
+
+from reconcile import queries
+from reconcile.gql_definitions.skupper_network.skupper_networks import SkupperNetworkV1
+from reconcile.gql_definitions.skupper_network.skupper_networks import (
+    query as skupper_networks_query,
+)
+from reconcile.skupper_network import reconciler
+from reconcile.skupper_network.models import (
+    SkupperConfig,
+    SkupperSite,
+)
+from reconcile.skupper_network.site_controller import CONFIG_NAME
+from reconcile.skupper_network.site_controller import LABELS as SITE_CONTROLLER_LABELS
+from reconcile.skupper_network.site_controller import (
+    site_config,
+    site_controller_deployment,
+    site_controller_role,
+    site_controller_role_binding,
+    site_controller_service_account,
+)
+from reconcile.utils import gql
+from reconcile.utils.defer import defer
+from reconcile.utils.disabled_integrations import integration_is_enabled
+from reconcile.utils.oc import OC_Map
+from reconcile.utils.openshift_resource import OpenshiftResource as OR
+from reconcile.utils.openshift_resource import ResourceInventory
+from reconcile.utils.semver_helper import make_semver
+
+QONTRACT_INTEGRATION = "skupper-network"
+QONTRACT_INTEGRATION_VERSION = make_semver(0, 1, 0)
+
+
+class SkupperNetworkExcpetion(Exception):
+    """Base exception for Skupper Network integration."""
+
+
+def compile_skupper_sites(
+    skupper_networks: Iterable[SkupperNetworkV1],
+) -> list[SkupperSite]:
+    """Compile the list of skupper sites to be created, updated or deleted."""
+    sites: list[SkupperSite] = []
+    for skupper_network in skupper_networks:
+        network_sites: list[SkupperSite] = []
+
+        for ns in skupper_network.namespaces:
+            if not ns.skupper_site:
+                # make mypy happy
+                continue
+
+            if not integration_is_enabled(QONTRACT_INTEGRATION, ns.cluster):
+                # integration is disabled for this cluster
+                continue
+
+            edge = ns.cluster.internal if ns.cluster.internal else False
```

Comment:
That is a good point; it was always on my mind. Referring to the [design doc](https://gitlab.cee.redhat.com/service/app-interface/-/blob/master/docs/app-sre/design-docs/skupper-integration.md#details), an internal cluster is behind the VPN, the question is, is it reachable by other internal clusters? Does it need a VPC peering? I still struggle with the `private` & `internal` attributes.

Reply:
hmmm, appsrep05ue1 is reachable from appsres05ue1 (and viceversa). I see no real reason to create a skupper link between them, though 😂

Reply:
@patjlm, any comment on that? Should an `internal` cluster always be an edge site?

Reply:
I would say that the more links we have the better, to obtain several routing opportunities. So `private` could connect to each other.
`internal` could be more debatable for security concerns, but since the traffic is bidirectional once a link is up, I don't really see the difference. So I'd agree a site on an internal cluster is not necessarily an edge. Maybe we should compute if there are any incoming link to know if we have an edge ?
Links should be created *only* between namespaces that are configured for that skupper network. If appsrep05ue1 and appsres03ue1 namespaces are not configured to have namespaces in the same network, they will not be linked. We could still have the need though, for prometheus access for example. So it is technically possible.

Reply:
That sounds reasonable to me.

Reply:
I've changed the connection algorithm; internal clusters aren't edge sites by default anymore.